### PR TITLE
Implement GetKey for Vec<Xpriv>

### PR DESCRIPTION
**Pull request by shesek**
It appears that the `BTreeSet<Xpriv>`/`HashSet<Xpriv>` sets currently implementing `GetKey` cannot actually be constructed, because `Xpriv` does not implement `Ord` nor `Hash`. (And that the rust-bitcoin code referencing these sets should not even compile? yet evidently it does :eyes: )

This PR adds support for `Vec<Xpriv>` to enable signing with multiple `Xpriv`s, but does not address the issue with the existing sets.

The added test case demonstrates the issue:

```rust
error[E0277]: the trait bound `bip32::Xpriv: std::hash::Hash` is not satisfied
    --> bitcoin/src/psbt/mod.rs:2301:24
     |
2301 |         HashSet::new().insert(xpriv.clone());
     |                        ^^^^^^ the trait `std::hash::Hash` is not implemented for `bip32::Xpriv`
     |
note: required by a bound in `std::collections::HashSet::<T, S>::insert`
    --> /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/collections/hash/set.rs:888:5

error[E0277]: the trait bound `bip32::Xpriv: Ord` is not satisfied
    --> bitcoin/src/psbt/mod.rs:2302:25
     |
2302 |         BTreeSet::new().insert(xpriv.clone());
     |                         ^^^^^^ the trait `Ord` is not implemented for `bip32::Xpriv`
     |
note: required by a bound in `std::collections::BTreeSet::<T, A>::insert`
    --> /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/collections/btree/set.rs:899:5

```

### CODE REVIEW

**Comment by apoelstra** [comment]
> It appears that the `BTreeSet<Xpriv>`/`HashSet<Xpriv>` sets currently implementing `GetKey` cannot actually be constructed, because `Xpriv` does not implement `Ord` nor `Hash`. (And that the rust-bitcoin code referencing these sets should not even compile? yet evidently it does 👀 )

I'm sure they can be constructed, but they likely cannot be populated or used :).

Anyway `Xpriv` should definitely implement `Hash` and probably impl `ArbitraryOrd` as well. We need to be a bit careful about constant-timeness here.

**Comment by storopoli** [comment]
> We need to be a bit careful about constant-timeness here.

What if we defer to `libsecp`? It has constant-timeness.

**Comment by apoelstra** [comment]
Thanks for adding the failing test BTW. I think the solution should be:

* Impl `Hash` for `Xpriv` (IMO we can just derive it, siphash won't leak timing info, but if you want to be cautious we should sha2 it first)
* Impl `ArbitraryOrd' for `Xpriv`
* Replace the `BTreeSet<Xpriv>` impl with `BTreeSet<Ordered<Xpriv>>`

**Comment by apoelstra** [comment]
> What if we defer to `libsecp`? It has constant-timeness.

It has constant-time equality/cmp (which yes, we should use, good call). This will be a little slow because of the FFI overhead but I think that's fine until somebody complains.

libsecp does not have a constant-time hash.

**Comment by Kixunil** [comment]
> * siphash won't leak timing info

`Hasher` is not guaranteed to be siphash, we should hash it ourselves.

If we can pull off constatnt time impl, then let's just do normal `Ord`.

**Comment by shesek** [comment]
Is there anything preventing adding support for `Vec`s?

**Review by Kixunil — APPROVED** [review]
ACK 0f6cfd8659cb31dc67095f5ec4cc2ba6da11c904

This has no constant-time issues because the key is already compared by `parent_fingerprint` which is a hash of *public* key.

**Comment by shesek** [comment]
@Kixunil Note that the current `GetKey` implementation for sets is incorrect -- the comparison should actually use the Xpriv own `fingerprint()` rather than its `master_fingerprint`, similarly to how its implemented for a single `Xpriv::get_key()` (see [here](https://github.com/rust-bitcoin/rust-bitcoin/blob/0c2737f25a2d2291633bd93e8137c51d1e82a569/bitcoin/src/psbt/mod.rs#L764) vs [here](https://github.com/rust-bitcoin/rust-bitcoin/blob/0c2737f25a2d2291633bd93e8137c51d1e82a569/bitcoin/src/psbt/mod.rs#L810)). This will internally call `Xpub::from_priv()` to compute the fingerprint.

I just recently created #3356 to address this.

**Comment by Kixunil** [comment]
Yes, I noticed it being weird but I had to run before I could respond to #3356

**Comment by apoelstra** [comment]
I'm not sure what to do with 0f6cfd8659cb31dc67095f5ec4cc2ba6da11c904 -- your "failing test case" is illustrative, but it fails by not compiling :). So we can't merge it as-is.

I am tempted to just deprecate the BTreeMap fields here and tell people to use the psbt crate (whose name I don't remember) whose API hopefully actually works.

**Comment by shesek** [comment]
> your "failing test case" is illustrative, but it fails by not compiling :)

I rebased on master, left the test but commented-out the failing `HashSet`/`BTreeSet` construction with a FIXME note. I can remove that commit entirely if that seems preferable, just thought it might be useful to have this mentioned as a heads up.

> I am tempted to just .. tell people to use the psbt crate

fwiw I've been playing with rust-bitcoin's (and rust-miniscirpt's) PSBT for awhile now, and my experience with it was pretty good overall. This PR plus #3356 were the only issues I ran into.

I have it working in Minsc, [here's an example](https://min.sc/v0.3/#c=%24alice%20%3D%20tprv8ZgxMBicQKsPdJ4fACcDyaRS9vuxRtBrj69BEficvR7UEk9DhyCVzQCYws8ZNqnY8U3C6vJ9xK8BPHjb9GGCXkdcyRqdF6Uy3gsGe4gjAgt%3B%0A%24bob%20%3D%20tprv8ZgxMBicQKsPdtLA8dMLSyajr6eKVyw2XpurTNH4vSsEouJbgEcg6yHctAGtdbu2N8JvpuUmT2mRHnhjuAQt7yDnCcBSqNAMqwUDGtyqhKR%3B%0A%0A%24psbt%20%3D%20psbt%20%5B%0A%20%20%22unsigned_tx%22%3A%20%5B%0A%20%20%20%20%22version%22%3A%202%2C%0A%20%20%20%20%22inputs%22%3A%20%5B%0A%20%20%20%20%20%2001ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b%3A0%2C%0A%20%20%20%20%20%20d5110331f105a29677a2f1b016c6db28e5fc1062650064ba926590d9a43862d3%3A2%2C%0A%20%20%20%20%20%20b09229665ef21aa1d66d44e9022ca66dc96c0bd3018b45cf4cc1efd5ae878a86%3A1%2C%0A%20%20%20%20%20%20%5B%0A%20%20%20%20%20%20%20%20%22prevout%22%3A%20386f2ba8c5b5b2f690337795d306d73e7f7e8167f7cdfadb1a5712f70b492a1f%3A5%2C%0A%20%20%20%20%20%20%20%20%22sequence%22%3A%201%20month%2C%0A%20%20%20%20%20%20%5D%0A%20%20%20%20%5D%2C%0A%20%20%20%20%22outputs%22%3A%20%5B%0A%20%20%20%20%20%20wpkh%28%24alice%2F1%2F4%29%3A1.5%20BTC%2C%0A%20%20%20%20%20%20tr%28%24bob%2F1%2F8%29%3A1.7%20BTC%2C%0A%20%20%20%20%20%20tb1qs7ucx8mz6c2p3845uqst0048hvvmqr709ruhe6%3A%200.1%20BTC%2C%0A%20%20%20%20%5D%0A%20%20%5D%2C%0A%20%20%22inputs%22%3A%20%5B%0A%20%20%20%20%2F%2F%20the%20utxo%20can%20be%20auto-constructed%20from%20the%20descriptor%20and%20utxo%20amount%0A%20%20%20%200%3A%20%5B%20%22descriptor%22%3A%20wpkh%28%24alice%2F0%2F100%29%2C%20%22amount%22%3A%200.3%20BTC%20%5D%2C%0A%20%20%20%201%3A%20%5B%20%22descriptor%22%3A%20tr%28%24bob%2F0%2F72%29%2C%20%22amount%22%3A%200.9%20BTC%20%5D%2C%0A%20%20%20%20%2F%2F%20alternatively%2C%20the%20descriptor%20can%20be%20extracted%20from%20the%20utxo%20%28if%20it%20has%20one%29%0A%20%20%20%202%3A%20%5B%20%22utxo%22%3A%20wsh%28pk%28%24bob%2F0%2F2%29%20%26%26%20pk%28%24alice%2F0%2F6%29%29%3A5%20BTC%20%5D%2C%0A%20%20%20%203%3A%20%5B%20%22utxo%22%3A%20tr%28pk%28%24bob%2F0%2F2%29%20%26%26%20older%281%20month%29%29%3A0.2%20BTC%20%5D%2C%0A%20%20%5D%2C%0A%5D%3B%0A%20%20%0Aassert%3A%3Aeq%28psbt%3A%3Afee%28%24psbt%29%2C%203.1%20BTC%29%3B%0A%0A%24tx%20%3D%20psbt%3A%3Aextract%28psbt%3A%3Afinalize%28psbt%3A%3Asign%28%24psbt%2C%20%5B%20%24alice%2C%20%24bob%20%5D%29%29%29%3B%0A) with some multisigs and timelocks. [Or here](https://min.sc/v0.3/#c=%24alice%20%3D%20tprv8ZgxMBicQKsPdJ4fACcDyaRS9vuxRtBrj69BEficvR7UEk9DhyCVzQCYws8ZNqnY8U3C6vJ9xK8BPHjb9GGCXkdcyRqdF6Uy3gsGe4gjAgt%3B%0A%24bob%20%3D%20tprv8ZgxMBicQKsPdtLA8dMLSyajr6eKVyw2XpurTNH4vSsEouJbgEcg6yHctAGtdbu2N8JvpuUmT2mRHnhjuAQt7yDnCcBSqNAMqwUDGtyqhKR%3B%0A%0A%24psbt%20%3D%20psbt%3A%3Atx%20%5B%0A%20%20%22version%22%3A%202%2C%0A%20%20%22inputs%22%3A%20%5B%0A%20%20%20%20%5B%0A%20%20%20%20%20%20%22prevout%22%3A%2001ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b%3A0%2C%0A%20%20%20%20%20%20%22utxo%22%3A%20wpkh%28%24alice%2F0%2F100%29%3A0.3%20BTC%2C%0A%20%20%20%20%5D%2C%0A%20%20%20%20%5B%0A%20%20%20%20%20%20%22prevout%22%3A%20d5110331f105a29677a2f1b016c6db28e5fc1062650064ba926590d9a43862d3%3A2%2C%0A%20%20%20%20%20%20%22utxo%22%3A%20tr%28%24bob%2F0%2F72%29%3A0.9%20BTC%2C%0A%20%20%20%20%5D%2C%0A%20%20%20%20%5B%0A%20%20%20%20%20%20%22prevout%22%3A%20b09229665ef21aa1d66d44e9022ca66dc96c0bd3018b45cf4cc1efd5ae878a86%3A1%2C%0A%20%20%20%20%20%20%22utxo%22%3A%20wsh%28pk%28%24bob%2F0%2F2%29%20%26%26%20pk%28%24alice%2F0%2F6%29%29%3A5%20BTC%2C%0A%20%20%20%20%5D%2C%0A%20%20%20%20%5B%0A%20%20%20%20%20%20%22prevout%22%3A%20386f2ba8c5b5b2f690337795d306d73e7f7e8167f7cdfadb1a5712f70b492a1f%3A5%2C%0A%20%20%20%20%20%20%22utxo%22%3A%20tr%28pk%28%24bob%2F0%2F2%29%20%26%26%20older%281%20month%29%29%3A0.2%20BTC%2C%0A%20%20%20%20%20%20%22sequence%22%3A%201%20month%2C%0A%20%20%20%20%5D%0A%20%20%5D%2C%0A%20%20%22outputs%22%3A%5B%0A%20%20%20%20wpkh%28%24alice%2F1%2F4%29%3A1.5%20BTC%2C%0A%20%20%20%20tr%28%24bob%2F1%2F8%29%3A1.7%20BTC%2C%0A%20%20%20%20wsh%28pk%28%24bob%2F8%29%20%26%26%20pk%28%24alice%2F8%29%29%3A0.1%20BTC%2C%0A%20%20%5D%0A%5D%3B%20%2F%2F%20all%20other%20PSBT%20fields%20can%20be%20set%20too%0A%0Aassert%3A%3Aeq%28psbt%3A%3Afee%28%24psbt%29%2C%203.1%20BTC%29%3B%0A%0A%24tx%20%3D%20psbt%3A%3Aextract%28psbt%3A%3Afinalize%28psbt%3A%3Asign%28%24psbt%2C%20%5B%20%24alice%2C%20%24bob%20%5D%29%29%29%3B) with an alternative API to construct the psbt+tx in one go (somewhat more similarly to PSBTv2, but uses PSBTv0).

**Comment by tcharding** [comment]
>  I can remove that commit entirely if that seems preferable

Yes please, we have made an effort recently to stick to github issues and not use `FIXME` or `TODO` in code. Just say if you don't have time to write the issue and I'll write it up.

Patch 1 looks good, thanks for the contribution!

**Comment by shesek** [comment]
> Yes please

Removed the test.

**Review by apoelstra — APPROVED** [review]
ACK 8ec3571d80e646e03b53d2eb46baaf8e9492879b successfully ran local tests

**Comment by tcharding** [comment]
Not really related to this PR but just FTR

> I am tempted to just deprecate the BTreeMap fields here and tell people to use the psbt crate (whose name I don't remember) whose API hopefully actually works.

If this refers to my `psbt-v2` [crate](https://crates.io/crates/psbt-v2) then the same bug is in it because it just imports the code from `psbt` module of release `v0.32.2` (found https://github.com/tcharding/rust-psbt/issues/32 while checking).

**Review by tcharding — APPROVED** [review]
ACK 8ec3571d80e646e03b53d2eb46baaf8e9492879b

**Pull request merged**
